### DISCUSSION TITLE
Todoの中身の値の受け渡し

### DISCUSSION THREAD

**Discussion by mahirooo** [lit-iphone-course/lit-iphone-course#14] [Question-original] [answered]
**やろうとしたこと：**
1.AddViewControllerのtodoとdetailを保存
2.保存したらViewControllerにtodoとdetailの値を渡す
3.todoとdetailの値を変数に入れる。
4.変数の値をtodoArrayとdetailArrayにそれぞれappendする
5.配列をCellに表示させる

**解決したいこと：**
値の受け渡しがうまくいっていないので、そこを解消したいです！

[Githubリンク→](https://github.com/mahirooo/OriginalTodo)

**Comment by txxxxc** [comment, marked as answer]
まず値の受け渡しがうまくいかない理由として、viewDidLoadが呼び出されるタイミングが関係しています！
viewDidLoadは初回の画面ロードの時限定で呼び出されるという性質があります。つまり、AddViewControllerから値を渡しても、viewDidLoadに書いてある配列に要素を追加する処理は呼び出されません！
```
// viewDidLoadは初回描画でしか発火しないので、以下のコードで配列に値は追加されない！
todoArray.append(todo)
detailArray.append(detail)
```

ですが、このような画面が表示されるたびに特定の処理を行いたいという目的のためのメソッドが用意されています！それが`viewWillAppear / viewDidApper`です！このメソッドは画面が表示されるたびに発火されるので、こういった場面でよく使われます！なので一回使ってみてください！（今回はviewWillAppearが良さそう）
新しくtodoArrayなどに値を追加する処理をviewWillAppearなどに書くと配列に追加されると思うので、試してみてください！
こういったviewDidLoadやviewWillAppearなどのメソッドのことをライフサイクルメソッドと呼びます！
参考は以下です！↓
https://ios-docs.dev/life-cycle/

注意点としては、画面遷移の方法がPresent Modallyの場合viewWillAppearが発火しないので、Presentationの方式をFullScreenにする必要があります！
![image](https://user-images.githubusercontent.com/22112440/188545077-bb6c01d8-1949-4df4-b76d-84354ba5fb87.png)

**その他の実装方法**
ちゃんぽんの実装方法でも十分実装することは可能ですが、知識として一般的な実装方法も知っていてほしいので一応記しておきます！

一般的な実装方法は以下の通りです！
1. TODO追加画面で追加したいTODOの情報を入力する
2. `追加`のボタンが押された時点でフォームに入力されたTODOをRealmに保存して、一覧画面に遷移させる。(画面を戻す)
3. 一覧画面で画面が再度表示されたことを検知(viewWillAppearなどを使って)し、Realmから保存されているTODOを再度すべて取り出し、テーブルに描画する

こっちのほうが楽そうとか思ったら全然こっちで実装してもらっても大丈夫です！
ご確認よろしくお願いしますー！
